### ISSUE TITLE
codeCoverage bugs seem to suggest API bugs.

### ISSUE THREAD

**Issue by ianamason** [dyninst/dyninst#586]
I was trying to see why the codeCoverage.dir example was crashing. 
If I don't pass the '-s' flag to `codeCoverage` nothing gets instrumented.
```
% ./codeCoverage -b ./testcc testcc.inst
```
If I do pass the '-s' flag
```
% ./codeCoverage -sb ./testcc testcc.inst
```
I seem to instrument way more than I want. It seems that the 
```
appImage->getModules ();
```
returns a vector of 1403 modules.  Every one claims to be a shared library.
Even the ones that obviously are not shared libraries such as: `testcc.c` 

The list includes lots of elements of libC which I thought we were skipping
because of the explicit  `skipLibraries` check.  In particular I see BOTH
`libInst.so` AND `libInst.C` in this list, and only the first will be skipped.

Can someone help?    I include the first 100 modules in the vector:

```
./testcc has 1403 modules
Module libInst.so is a shared library
Module libInst.C is a shared library
Module testcc is a shared library
Module testcc.c is a shared library
Module libtestcc.so is a shared library
Module libtestcc.c is a shared library
Module rtld.c is a shared library
Module ld-linux-x86-64.so.2 is a shared library
Module dl-tls.c is a shared library
Module dl-load.c is a shared library
Module dl-lookup.c is a shared library
Module dl-object.c is a shared library
Module dl-reloc.c is a shared library
Module dl-deps.c is a shared library
Module dl-hwcaps.c is a shared library
Module dl-runtime.c is a shared library
Module dl-init.c is a shared library
Module dl-fini.c is a shared library
Module dl-debug.c is a shared library
Module dl-misc.c is a shared library
Module dl-version.c is a shared library
Module dl-profile.c is a shared library
Module dl-origin.c is a shared library
Module dl-scope.c is a shared library
Module dl-execstack.c is a shared library
Module dl-caller.c is a shared library
Module dl-open.c is a shared library
Module dl-close.c is a shared library
Module dl-trampoline.h is a shared library
Module dl-exception.c is a shared library
Module dl-sort-maps.c is a shared library
Module dl-cache.c is a shared library
Module dl-tunables.c is a shared library
Module tlsdesc.c is a shared library
Module dl-tlsdesc.S is a shared library
Module tls_get_addr.S is a shared library
Module dl-get-cpu-features.c is a shared library
Module dl-sysdep.c is a shared library
Module dl-environ.c is a shared library
Module dl-minimal.c is a shared library
Module dl-error-minimal.c is a shared library
Module dl-conflict.c is a shared library
Module dl-brk.c is a shared library
Module dl-sbrk.c is a shared library
Module dl-getcwd.c is a shared library
Module dl-openat64.c is a shared library
Module dl-opendir.c is a shared library
Module dl-fxstatat64.c is a shared library
Module check_fds.c is a shared library
Module closedir.c is a shared library
Module readdir.c is a shared library
Module rewinddir.c is a shared library
Module getdents.c is a shared library
Module fdopendir.c is a shared library
Module dl-addr-obj.c is a shared library
Module profil.c is a shared library
Module prof-freq.c is a shared library
Module xstat.c is a shared library
Module fxstat.c is a shared library
Module lxstat.c is a shared library
Module open64.c is a shared library
Module openat64.c is a shared library
Module read.c is a shared library
Module write.c is a shared library
Module lseek64.c is a shared library
Module access.c is a shared library
Module fcntl.c is a shared library
Module close.c is a shared library
Module mmap64.c is a shared library
Module syscall-template.S is a shared library
Module _exit.c is a shared library
Module setjmp.S is a shared library
Module __longjmp.S is a shared library
Module sigaction.c is a shared library
Module strdup.c is a shared library
Module strcmp.S is a shared library
Module memchr.S is a shared library
Module rawmemchr.S is a shared library
Module memcmp.S is a shared library
Module strchr.S is a shared library
Module strlen.S is a shared library
Module strcpy.S is a shared library
Module strcspn.S is a shared library
Module memmove-vec-unaligned-erms.S is a shared library
Module memset-vec-unaligned-erms.S is a shared library
Module libc.so.6 is a shared library
Module unwind-resume.c is a shared library
Module gconv_db.c is a shared library
Module libc_fatal.c is a shared library
Module malloc.c is a shared library
Module regex.c is a shared library
Module resolv_conf.c is a shared library
Module cacheinfo.c is a shared library
Module init-first.c is a shared library
Module libc-start.c is a shared library
Module version.c is a shared library
Module check_fds.c is a shared library
Module errno-loc.c is a shared library
.
.
.
```

**Comment by ianamason** [comment]
I note that if I flick the "openDependencies" to `false` in the `openBinary` call I get a simpler 
picture, but they still ALL claim to be shared libraries.
```
./testcc has 4 modules
Module libInst.so is a shared library
Module libInst.C is a shared library
Module testcc is a shared library
Module testcc.c is a shared library
```

**Comment by mxz297** [comment]
> I was trying to see why the codeCoverage.dir example was crashing.
> If I don't pass the '-s' flag to `codeCoverage` nothing gets instrumented.

I cannot reproduce your problem. I just ran codeCoverage as what you did and got the following output:
```
[xmeng@mic codeCoverage.dir]$ ~/dyninstapi/install/bin/codeCoverage -b ./testcc testcc.inst
Instrumenting module: testcc
Inserting instrumention at function entry of _init
Instrumenting Basic Block 0x4004a0 of _init
Instrumenting Basic Block 0x4004b0 of _init
Instrumenting Basic Block 0x4004b2 of _init
Inserting instrumention at function entry of _start
Instrumenting Basic Block 0x4004e0 of _start
Instrumenting Basic Block 0x40050a of _start
Inserting instrumention at function entry of __do_global_dtors_aux
Instrumenting Basic Block 0x400580 of __do_global_dtors_aux
Instrumenting Basic Block 0x400589 of __do_global_dtors_aux
Instrumenting Basic Block 0x400592 of __do_global_dtors_aux
Instrumenting Basic Block 0x4005a0 of __do_global_dtors_aux
Inserting instrumention at function entry of frame_dummy
Instrumenting Basic Block 0x4005b0 of frame_dummy
Inserting instrumention at function entry of three
Instrumenting Basic Block 0x4005b7 of three
Inserting instrumention at function entry of two
Instrumenting Basic Block 0x4005c8 of two
Instrumenting Basic Block 0x4005d9 of two
Instrumenting Basic Block 0x4005e3 of two
Inserting instrumention at function entry of one
Instrumenting Basic Block 0x4005f0 of one
Instrumenting Basic Block 0x400608 of one
Instrumenting Basic Block 0x40060e of one
Instrumenting Basic Block 0x400618 of one
Instrumenting Basic Block 0x400619 of one
Inserting instrumention at function entry of main
Instrumenting Basic Block 0x40061b of main
Instrumenting Basic Block 0x40063e of main
Instrumenting Basic Block 0x400645 of main
Instrumenting Basic Block 0x40064b of main
Instrumenting Basic Block 0x400652 of main
Instrumenting Basic Block 0x400661 of main
Inserting instrumention at function entry of __libc_csu_init
Instrumenting Basic Block 0x400670 of __libc_csu_init
Instrumenting Basic Block 0x4006a1 of __libc_csu_init
Instrumenting Basic Block 0x4006a6 of __libc_csu_init
Instrumenting Basic Block 0x4006b0 of __libc_csu_init
Instrumenting Basic Block 0x4006bd of __libc_csu_init
Instrumenting Basic Block 0x4006c6 of __libc_csu_init
Inserting instrumention at function entry of __libc_csu_fini
Instrumenting Basic Block 0x4006e0 of __libc_csu_fini
Inserting instrumention at function entry of _fini
Instrumenting Basic Block 0x4006e4 of _fini
insertInitCallback on testcc
insertFiniCallback on testcc
[xmeng@mic codeCoverage.dir]$ ./testcc.inst 


 ************************** Code Coverage ************************* 

    1 : __do_global_dtors_aux, testcc
    1 : _fini, testcc
    1 : _init, testcc
    1 : frame_dummy, testcc
    1 : main, testcc
    1 : one, testcc
    1 : two, testcc

 ************** Code Coverage 7 out of 11 functions ************** 



 ************************** Basic Block Coverage ************************* 

 (__do_global_dtors_aux, testcc)
 	    1 : 0x400580  
 	    1 : 0x400589  
 	    1 : 0x400592  
 (__libc_csu_init, testcc)
 	    1 : 0x4006c6  
 	    1 : 0x4006a6  
 	    1 : 0x4006b0  
 	    1 : 0x4006bd  
 	    1 : 0x4006a1  
 (_fini, testcc)
 	    1 : 0x4006e4  
 (_init, testcc)
 	    1 : 0x4004b2  
 	    1 : 0x4004a0  
 (frame_dummy, testcc)
 	    1 : 0x4005b0  
 (main, testcc)
 	    1 : 0x400652  
 	    1 : 0x400661  
 	    1 : 0x400645  
 	    1 : 0x40061b  
 (one, testcc)
 	    1 : 0x4005f0  
 	    1 : 0x400608  
 	    1 : 0x400618  
 	    1 : 0x400619  
 (two, testcc)
 	    1 : 0x4005e3  
 	    1 : 0x4005c8  

 ************** Basic Block Coverage 22 out of 33 blocks ************** 
```

What is your architecture, linux distribution, compiler version? 


> ```
> appImage->getModules ();
> ```
> returns a vector of 1403 modules. Every one claims to be a shared library.
> Even the ones that obviously are not shared libraries such as: `testcc.c`

I am not sure why you got the impression that `BPatch_module` corresponds to shared library. `BPatch_object` corresponds to an a.out or a shared library. `BPatch_module` corresponds to a compilation unit. 

Can you tell us how you got the wrong information? We should fix such misleading information.

**Comment by ianamason** [comment]
I am working on a clean ubuntu 18.04 vagrant box.  I think you misunderstand what I meant 
about being a shared library. I meant that the modules `isSharedLib` returns `true`.
So  I never instrument ` testcc` because it claims to be a shared library, and so is skipped
in the codeCoverage module loop.

**Comment by hainest** [comment]
@ianamason could you post the output of `gcc -v`?

**Comment by mxz297** [comment]
I see. Does your compiler by default to generate PIE code?

**Comment by ianamason** [comment]
```
vagrant@ubuntu-bionic:/vagrant/Horus/examples/codeCoverage.dir$ gcc --version
gcc (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

vagrant@ubuntu-bionic:/vagrant/Horus/examples/codeCoverage.dir$ ./codeCoverage -b testcc testcc.inst
vagrant@ubuntu-bionic:/vagrant/Horus/examples/codeCoverage.dir$ ./testcc.inst
vagrant@ubuntu-bionic:/vagrant/Horus/examples/codeCoverage.dir$
```

**Comment by hainest** [comment]
@ianamason The stuff I am curious about is in `gcc -v`, not `gcc --version`. As @mxz297 noted, you could be generating PIE code. This is the default on Ubuntu 18.04 with gcc-8. I was curious if your compiler was also built with `--default-enable-pie`.

**Comment by ianamason** [comment]
Sorry
```
 gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/7/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 7.3.0-27ubuntu1~18.04' --with-bugurl=file:///usr/share/doc/gcc-7/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++ --prefix=/usr --with-gcc-major-version-only --program-suffix=-7 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --enable-default-pie --with-system-zlib --with-target-system-zlib --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 7.3.0 (Ubuntu 7.3.0-27ubuntu1~18.04)

```

**Comment by ianamason** [comment]
So yes. It looks to be generating PIE by default.

**Comment by hainest** [comment]
>--enable-default-pie

Yup. You are building PIE code by default. Try adding '-fno-pie' to the compile commands. I have a feeling Dyninst might be confusing PIE and PIC and just saying they are both shared library code (which isn't entirely inaccurate, but we need to endeavor to do better).

**Comment by ianamason** [comment]
@hainest  just the mutatee compiles, right?  It doesn't seem to fix the `isSharedLib`  but does alter the outcome (but not in a good way):
```
vagrant@ubuntu-bionic:/vagrant/Horus/examples/codeCoverage.dir$ make
g++ -g -Wall -I/home/vagrant/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.3.0/dyninst-10.0.0-puoloud2b2nvnfbvbbkwt2y442paycvb/include -I/home/vagrant/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.3.0/boost-1.69.0-bhagolkqgwcr3csckw2msh46me3zw3fl/include -I/home/vagrant/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.3.0/intel-tbb-2019.4-x7jmqru4rcixiq56llddow2qw5wca7uz/include -c codeCoverage.C
codeCoverage.C: In function ‘int main(int, char**)’:
codeCoverage.C:290:20: warning: variable ‘defaultModule’ set but not used [-Wunused-but-set-variable]
     BPatch_module *defaultModule;
                    ^~~~~~~~~~~~~
g++ -g -Wall  -I/home/vagrant/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.3.0/boost-1.69.0-bhagolkqgwcr3csckw2msh46me3zw3fl/include -L/home/vagrant/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.3.0/dyninst-10.0.0-puoloud2b2nvnfbvbbkwt2y442paycvb/lib \
-o codeCoverage codeCoverage.o  \
-L/home/vagrant/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.3.0/libiberty-2.31.1-6kikjqikkfawmkzvwxrykob5s5qzjhw7/lib -lcommon -liberty -ldyninstAPI -lpthread
g++ -g -Wall -fpic -shared libInst.C -o libInst.so
gcc -Wall -pedantic -g -std=gnu99 -fno-pie -fpic -shared -o libtestcc.so libtestcc.c
gcc -Wall -pedantic -g -std=gnu99 -fno-pie  -c testcc.c
gcc -Wall -pedantic -g -std=gnu99 -fno-pie -o testcc testcc.o ./libtestcc.so
vagrant@ubuntu-bionic:/vagrant/Horus/examples/codeCoverage.dir$ ./codeCoverage -b testcc testcc.inst
vagrant@ubuntu-bionic:/vagrant/Horus/examples/codeCoverage.dir$ ./testcc.inst
./testcc.inst: Symbol `libFooFunction' causes overflow in R_X86_64_PC32 relocation
vagrant@ubuntu-bionic:/vagrant/Horus/examples/codeCoverage.dir$
```

**Comment by hainest** [comment]
I was afraid of that. You will need to use the HEAD of the Dyninst master branch to get the necessary fixes for PIC code. They will be included in the 10.1 release which we are diligently working hard to get out soon.

**Comment by ianamason** [comment]
OK my spack-fu is not great. I guess I can't use spack to install the head right? I

**Comment by mxz297** [comment]
Can you try one more time where fno-pie is only applied to the executable, but not the .so?

**Comment by ianamason** [comment]
Same outcome, no instrumentation, and an overflow.
```
g++ -g -Wall -fpic -shared libInst.C -o libInst.so
gcc -Wall -pedantic -g -std=gnu99 -fpic -shared -o libtestcc.so libtestcc.c
gcc -Wall -pedantic -g -std=gnu99 -fno-pie  -c testcc.c
gcc -Wall -pedantic -g -std=gnu99 -fno-pie -o testcc testcc.o ./libtestcc.so
vagrant@ubuntu-bionic:/vagrant/Horus/examples/codeCoverage.dir$ ./codeCoverage -b testcc testcc.inst
vagrant@ubuntu-bionic:/vagrant/Horus/examples/codeCoverage.dir$ ./codeCoverage -b testcc testcc.inst
vagrant@ubuntu-bionic:/vagrant/Horus/examples/codeCoverage.dir$ ./testcc.inst
./testcc.inst: Symbol `libFooFunction' causes overflow in R_X86_64_PC32 relocation
```

**Comment by mxz297** [comment]
Closed via #605
